android/KMEA/app/src/main/java/com/tavultesoft/kmea/logic/ResourcesUpdateTool.java

```diff
@@ -0,0 +1,492 @@
+      channel.setDescription(description);
+      // Register the channel with the system; you can't change the importance
+      // or other notification behaviors after this
+      NotificationManager notificationManager = aContext.getSystemService(NotificationManager.class);
+      notificationManager.createNotificationChannel(channel);
+    }
+  }
+
+  public void onUpdateDetection(final List<Bundle> updatableResources) {
+    failedUpdateCount = 0;
+
+    if (USE_UPDATE_INDICATOR_ONLY)
+    {
+      for (Bundle resourceBundle : updatableResources) {
+        addOpenUpdate( resourceBundle);
+      }
+    }
+    else if(SEND_UPDATE_NOTIFICATIONS) {
+      for (Bundle resourceBundle : updatableResources) {
+        sendNotification( resourceBundle);
+      }
+    }
+    else {
+      AlertDialog.Builder dialogBuilder = new AlertDialog.Builder(currentContext);
+      dialogBuilder.setTitle(currentContext.getString(R.string.keyboard_updates_available));
+      dialogBuilder.setMessage(currentContext.getString(R.string.confirm_update));
+      dialogBuilder.setPositiveButton(currentContext.getString(R.string.label_update), new DialogInterface.OnClickListener() {
+        public void onClick(DialogInterface dialog, int which) {
+          // Update keyboards
+          if (KMManager.hasConnection(currentContext)) {
+            // For each updatable keyboard, one at a time, do the update.
+            // TODO:  May need a reework to better handle large amounts of updates -
+            //        these calls will stack within the Android subsystem and may have issues accordingly.
+            for (Bundle resourceBundle : updatableResources) {
+
+             addOpenUpdate(resourceBundle);
+
+              Intent intent = new Intent(currentContext, KMKeyboardDownloaderActivity.class);
+              intent.putExtras(resourceBundle);
+              currentContext.startActivity(intent);
+            }
+          } else {
+            Toast.makeText(currentContext, "No internet connection", Toast.LENGTH_SHORT).show();
+            checkingUpdates = false;
+          }
+        }
+      });
+
+      dialogBuilder.setNegativeButton(currentContext.getString(R.string.label_later), new DialogInterface.OnClickListener() {
+        public void onClick(DialogInterface dialog, int which) {
+          lastUpdateCheck = Calendar.getInstance();
+          checkingUpdates = false;
+        }
+      });
+
+      AlertDialog dialog = dialogBuilder.create();
+      if (isContextAvailable()) {
+        dialog.setOnCancelListener(new DialogInterface.OnCancelListener() {
+          @Override
+          public void onCancel(DialogInterface dialog) {
+            lastUpdateCheck = Calendar.getInstance();
+            checkingUpdates = false;
+          }
+        });
+        dialog.show();
+      } else {
+        checkingUpdates = false;
+      }
+    }
+  }
+
+  /**
+   * Add open update.
```

Comment:
Can you remind me what an "open update" is?
I see it frequently in this section.

Reply:
@darcywong00: open update list contains all updates detected by the update check but not executed/finished yet.